Convert streams to async.

clokep opened the pull request:
This is a tiny PR, but it was standalone so I figure it was worth putting up.

Notable that most of these functions weren't returning a `Deferred` anymore, it seems only the `RoomEventSource.get_current_key()` does. The rest just return integers.

anoadragon453 commented:
> Notable that most of these functions weren't returning a Deferred anymore, it seems only the RoomEventSource.get_current_key() does. The rest just return integers.

This seems a little confusing. Should we just make `get_current_key` `async` like we did with `_serialize_payload` in https://github.com/matrix-org/synapse/pull/7987?

erikjohnston commented:
> > Notable that most of these functions weren't returning a Deferred anymore, it seems only the RoomEventSource.get_current_key() does. The rest just return integers.
> 
> This seems like it could lead to confusion. Should we just make `get_current_key` `async` like we did with `_serialize_payload` in #7987?

I think if you follow the call flow all the way down you eventually reach a non-async function, so I would say no. The call stack for getting the current stream position for events is unnecessarily complicated and needlessly jumps through ten layers, but it should be fairly easy to simplify down to a single sync call.

clokep commented:
I added a couple more changes:
* Converted `get_room_events_max_id` to async/await.
* Make `RoomEventSource.get_current_key` sync and then propagate this back to `EventSources.get_current_token` and `EventSources.get_current_token_for_pagination`.
